### ISSUE TITLE
SystemError: <class 'numpy.linalg.LinAlgError'> returned a result with an exception set

### ISSUE THREAD

**Issue by MrOlm** [MaayanLab/blitzgsea#12]
Hello-

It seems that sometimes, more or less randomly, I'm hitting the error "SystemError: <class 'numpy.linalg.LinAlgError'> returned a result with an exception set".

I've uploaded some test data in the zip file below, and the python code I use to reproduce the error.

Thanks again for the great tool and for being so responsive!
Matt

Code using uploaded test data:
```
import blitzgsea as blitz
import pickle

with open('/home/mattolm/user_data/scratch/gsea_files/GUT_GENOME014039.fna.tdb.pickle', 'rb') as handle:
    tdb1 = pickle.load(handle)
with open('/home/mattolm/user_data/scratch/gsea_files/GUT_GENOME042029.fna.tdb.pickle', 'rb') as handle:
    tdb2 = pickle.load(handle)
with open('/home/mattolm/user_data/scratch/gsea_files/k2p.pickle', 'rb') as handle:
    k2p = pickle.load(handle)
    
# works
rdb = blitz.gsea(tdb1, k2p)
print(rdb.sort_values('fdr').head())

# fails
rdb = blitz.gsea(tdb2, k2p)
print(rdb.sort_values('fdr'))
```

Error traceback:
```
/home/mattolm/miniconda3/envs/mamba/lib/python3.10/site-packages/scipy/stats/_continuous_distns.py:3132: RuntimeWarning: invalid value encountered in double_scalars
  func = lambda a: np.log(a) - sc.digamma(a) - s
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
ValueError: On entry to DLASCL parameter number 4 had an illegal value

The above exception was the direct cause of the following exception:

SystemError                               Traceback (most recent call last)
Cell In[23], line 13
     10 rdb = blitz.gsea(tdb1, k2p)
     11 print(rdb.sort_values('fdr').head())
---> 13 rdb = blitz.gsea(tdb2, k2p)
     14 print(rdb.sort_values('fdr'))

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/blitzgsea/__init__.py:291, in gsea(signature, library, permutations, anchors, min_size, max_size, processes, plotting, verbose, progress, symmetric, signature_cache, kl_threshold, kl_bins, shared_null, seed, add_noise)
    289     f_alpha_pos, f_beta_pos, f_pos_ratio, ks_pos, ks_neg = pdf_cache[sig_hash]["model"]
    290 else:
--> 291     f_alpha_pos, f_beta_pos, f_pos_ratio, ks_pos, ks_neg = estimate_parameters(signature, abs_signature, signature_map, library, permutations=permutations, calibration_anchors=anchors, processes=processes, symmetric=symmetric, plotting=plotting, verbose=verbose, seed=seed, progress=progress, max_size=max_size)
    292     pdf_cache[sig_hash] = {}
    293     xv, pdf = create_pdf(np.array(signature["v"]), kl_bins)

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/blitzgsea/__init__.py:147, in estimate_parameters(signature, abs_signature, signature_map, library, permutations, max_size, symmetric, calibration_anchors, plotting, processes, verbose, progress, seed)
    143     print('Significant unbalance between positive and negative enrichment scores detected. Signature values are not centered close to 0.')
    145 x = np.array(x, dtype=float)
--> 147 f_alpha_pos = loess_interpolation(x, alpha_pos)
    148 f_beta_pos = loess_interpolation(x, beta_pos, frac=0.2)
    150 f_alpha_neg = loess_interpolation(x, alpha_neg)

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/blitzgsea/__init__.py:105, in loess_interpolation(x, y, frac)
    103 def loess_interpolation(x, y, frac=0.5):
    104     yl = np.array(y)
--> 105     xout, yout, wout = loess_1d(x, yl, frac=frac)
    106     return interpolate.interp1d(xout, yout)

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/loess/loess_1d.py:276, in loess_1d(x, y, xnew, degree, frac, npoints, rotate, sigy)
    274 biweights = (1 - uu)**2
    275 tot_weights = dist_weights*biweights
--> 276 poly = polyfit1d(x[w], y[w], degree, tot_weights)
    277 yfit = poly.yfit
    278 badOld = bad

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/loess/loess_1d.py:83, in polyfit1d.__init__(self, x, y, degree, weights)
     81 a = x[:, None]**np.arange(degree + 1)
     82 self.degree = degree
---> 83 self.coeff = np.linalg.lstsq(a*sqw[:, None], y*sqw, rcond=None)[0]
     84 self.yfit = a @ self.coeff

File <__array_function__ internals>:180, in lstsq(*args, **kwargs)

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/numpy/linalg/linalg.py:2300, in lstsq(a, b, rcond)
   2297 if n_rhs == 0:
   2298     # lapack can't handle n_rhs = 0 - so allocate the array one larger in that axis
   2299     b = zeros(b.shape[:-2] + (m, n_rhs + 1), dtype=b.dtype)
-> 2300 x, resids, rank, s = gufunc(a, b, rcond, signature=signature, extobj=extobj)
   2301 if m == 0:
   2302     x[...] = 0

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/numpy/linalg/linalg.py:101, in _raise_linalgerror_lstsq(err, flag)
    100 def _raise_linalgerror_lstsq(err, flag):
--> 101     raise LinAlgError("SVD did not converge in Linear Least Squares")

SystemError: <class 'numpy.linalg.LinAlgError'> returned a result with an exception set
```

[gseadata.zip](https://github.com/MaayanLab/blitzgsea/files/11376067/gseadata.zip)

**Comment by MrOlm** [comment]
I updated my dependencies and am now hitting a different error. I am providing a list of the dependencies as well as an attachment.

Code:
```
import blitzgsea as blitz
import pickle
import pandas as pd
tdb1 = pd.read_pickle('/home/mat
tdb2 = pd.read_pickle('/home/mat
with open('/home/mattolm/user_da
    k2p = pickle.load(handle)
    
rdb = blitz.gsea(tdb1, k2p)
print(rdb.sort_values('fdr').hea
rdb = blitz.gsea(tdb2, k2p)
print(rdb.sort_values('fdr'))
```

Error:
```
/home/mattolm/miniconda3/envs/mamba/lib/python3.10/site-packages/scipy/stats/_continuous_distns.py:3132: RuntimeWarning: invalid value encountered in scalar subtract
  func = lambda a: np.log(a) - sc.digamma(a) - s
---------------------------------------------------------------------------
LinAlgError                               Traceback (most recent call last)
Cell In[1], line 14
     11 rdb = blitz.gsea(tdb1, k2p)
     12 print(rdb.sort_values('fdr').head())
---> 14 rdb = blitz.gsea(tdb2, k2p)
     15 print(rdb.sort_values('fdr'))

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/blitzgsea/__init__.py:291, in gsea(signature, library, permutations, anchors, min_size, max_size, processes, plotting, verbose, progress, symmetric, signature_cache, kl_threshold, kl_bins, shared_null, seed, add_noise)
    289     f_alpha_pos, f_beta_pos, f_pos_ratio, ks_pos, ks_neg = pdf_cache[sig_hash]["model"]
    290 else:
--> 291     f_alpha_pos, f_beta_pos, f_pos_ratio, ks_pos, ks_neg = estimate_parameters(signature, abs_signature, signature_map, library, permutations=permutations, calibration_anchors=anchors, processes=processes, symmetric=symmetric, plotting=plotting, verbose=verbose, seed=seed, progress=progress, max_size=max_size)
    292     pdf_cache[sig_hash] = {}
    293     xv, pdf = create_pdf(np.array(signature["v"]), kl_bins)

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/blitzgsea/__init__.py:147, in estimate_parameters(signature, abs_signature, signature_map, library, permutations, max_size, symmetric, calibration_anchors, plotting, processes, verbose, progress, seed)
    143     print('Significant unbalance between positive and negative enrichment scores detected. Signature values are not centered close to 0.')
    145 x = np.array(x, dtype=float)
--> 147 f_alpha_pos = loess_interpolation(x, alpha_pos)
    148 f_beta_pos = loess_interpolation(x, beta_pos, frac=0.2)
    150 f_alpha_neg = loess_interpolation(x, alpha_neg)

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/blitzgsea/__init__.py:105, in loess_interpolation(x, y, frac)
    103 def loess_interpolation(x, y, frac=0.5):
    104     yl = np.array(y)
--> 105     xout, yout, wout = loess_1d(x, yl, frac=frac)
    106     return interpolate.interp1d(xout, yout)

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/loess/loess_1d.py:276, in loess_1d(x, y, xnew, degree, frac, npoints, rotate, sigy)
    274 biweights = (1 - uu)**2
    275 tot_weights = dist_weights*biweights
--> 276 poly = polyfit1d(x[w], y[w], degree, tot_weights)
    277 yfit = poly.yfit
    278 badOld = bad

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/loess/loess_1d.py:83, in polyfit1d.__init__(self, x, y, degree, weights)
     81 a = x[:, None]**np.arange(degree + 1)
     82 self.degree = degree
---> 83 self.coeff = np.linalg.lstsq(a*sqw[:, None], y*sqw, rcond=None)[0]
     84 self.yfit = a @ self.coeff

File <__array_function__ internals>:200, in lstsq(*args, **kwargs)

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/numpy/linalg/linalg.py:2285, in lstsq(a, b, rcond)
   2282 if n_rhs == 0:
   2283     # lapack can't handle n_rhs = 0 - so allocate the array one larger in that axis
   2284     b = zeros(b.shape[:-2] + (m, n_rhs + 1), dtype=b.dtype)
-> 2285 x, resids, rank, s = gufunc(a, b, rcond, signature=signature, extobj=extobj)
   2286 if m == 0:
   2287     x[...] = 0

File ~/miniconda3/envs/mamba/lib/python3.10/site-packages/numpy/linalg/linalg.py:101, in _raise_linalgerror_lstsq(err, flag)
    100 def _raise_linalgerror_lstsq(err, flag):
--> 101     raise LinAlgError("SVD did not converge in Linear Least Squares")

LinAlgError: SVD did not converge in Linear Least Squares
```


[deps.txt](https://github.com/MaayanLab/blitzgsea/files/11376394/deps.txt)

**Comment by lachmann12** [comment]
Ok, this is strange. I have now tested it on two systems. One on Python 3.9 and one on python 3.10. 

Just to avoid any randomness I added a seed of 1:

import blitzgsea as blitz
import pandas as pd

tdb1 = pd.read_pickle('data/GUT_GENOME014039.fna.tdb.pickle')
tdb2 = pd.read_pickle('data/GUT_GENOME042029.fna.tdb.pickle')
k2p = pd.read_pickle('data/k2p.pickle')

rdb1 = blitz.gsea(tdb1, k2p, seed=1)
rdb2 = blitz.gsea(tdb2, k2p, seed=1)

This seems to work for me, and the dependencies seem the same as well. There could be a bug in the underlying c libraries that numpy uses. Specifically BLAS. 

It is clear that the error is triggered by function loess_1d. I have replaced the loess library. Let's hope this is the solution. 

Let me know if it works now before i close this issue.
